pydantic/dataclasses.py

```diff
@@ -264,6 +265,35 @@ def create_dataclass(cls: type[Any]) -> type[PydanticDataclass]:
             **kwargs,
         )
 
+        if config_wrapper.validate_assignment:
+
+            @functools.wraps(cls.__setattr__)
+            def validated_setattr(instance: Any, field: str, value: str, /) -> None:
+                type(instance).__pydantic_validator__.validate_assignment(instance, field, value)
+
+            cls.__setattr__ = validated_setattr.__get__(None, cls)  # type: ignore
+
+            if slots and not hasattr(cls, '__setstate__'):
+                # If slots is set, `pickle` (relied on by `copy.copy()`) will use
+                # `__setattr__()` to reconstruct the dataclass. However, the custom
+                # `__setattr__()` set above relies on `validate_assignment()`, which
+                # in turn excepts all the field values to be already present on the
+                # instance, resulting in attribute errors.
+                # As such, we make use of `object.__setattr__()` instead.
+                # Note that we do so only if `__setstate__()` isn't already set (this is the
+                # case if on top of `slots`, `frozen` is used).
+
+                # Taken from `dataclasses._dataclass_get/setstate()`:
+                def _dataclass_getstate(self: Any) -> list[Any]:
+                    return [getattr(self, f.name) for f in dataclasses.fields(self)]
+
+                def _dataclass_setstate(self: Any, state: list[Any]) -> None:
+                    for field, value in zip(dataclasses.fields(self), state):
+                        object.__setattr__(self, field.name, value)
```

Comment:
Should we run validation after this?

Reply:
Pickling is used with the assumption that the data is trusted already (if you need validation to be performed, then you have bigger problems as you are most likely vulnerable to arbitrary code execution). See `BaseModel.__setstate__()` as well for reference where we don't apply validation.